### PR TITLE
feat: add versioning to transaction(input,output,kernel)

### PR DESCRIPTION
**Pull request by Cifko**
Description
---
Add versioning to Transactions.

How Has This Been Tested?
---
Some cucumber tests.

### CODE REVIEW

File: base_layer/core/src/transactions/transaction/output_features.rs [hunk 1]
```diff
@@ -43,9 +44,26 @@ use crate::{
...
+impl TryFrom<u8> for OutputFeaturesVersion {
+    type Error = String;
+
+    fn try_from(value: u8) -> Result<Self, Self::Error> {
+        match value {
+            0 => Ok(OutputFeaturesVersion::DIBBLER),
+            _ => Err("Unknown version!".to_string()),
+        }
+    }
+}
```
Comment (hansieodendaal) [inline base_layer/core/src/transactions/transaction/output_features.rs]: I think this should go in a separate file

File: base_layer/core/src/transactions/transaction/transaction_input.rs [hunk 2]
```diff
@@ -56,11 +58,29 @@ use crate::{
...
+    type Error = String;
+
+    fn try_from(value: u8) -> Result<Self, Self::Error> {
+        match value {
+            0 => Ok(TransactionInputVersion::DIBBLER),
+            _ => Err("Unknown version!".to_string()),
+        }
+    }
+}
+
```
Comment (hansieodendaal) [inline base_layer/core/src/transactions/transaction/transaction_input.rs]: I think this should go in a separate file

File: base_layer/core/src/transactions/transaction/transaction_kernel.rs [hunk 3]
```diff
@@ -39,13 +40,30 @@ use crate::transactions::{
...
+impl TryFrom<u8> for TransactionKernelVersion {
+    type Error = String;
+
+    fn try_from(value: u8) -> Result<Self, Self::Error> {
+        match value {
+            0 => Ok(TransactionKernelVersion::DIBBLER),
+            _ => Err("Unknown version!".to_string()),
+        }
+    }
+}
```
Comment (hansieodendaal) [inline base_layer/core/src/transactions/transaction/transaction_kernel.rs]: I think this should go in a separate file

File: base_layer/core/src/transactions/transaction/transaction_output.rs [hunk 4]
```diff
@@ -70,11 +71,29 @@ use crate::{
...
+    type Error = String;
+
+    fn try_from(value: u8) -> Result<Self, Self::Error> {
+        match value {
+            0 => Ok(TransactionOutputVersion::DIBBLER),
+            _ => Err("Unknown version!".to_string()),
+        }
+    }
+}
+
```
Comment (hansieodendaal) [inline base_layer/core/src/transactions/transaction/transaction_output.rs]: I think this should go in a separate file

File: base_layer/wallet/src/output_manager_service/storage/sqlite_db/mod.rs [hunk 5]
```diff
@@ -756,8 +756,10 @@ impl OutputManagerBackend for OutputManagerSqliteDatabase {
         let start = Instant::now();
         let conn = self.database_connection.get_pooled_connection()?;
         let acquire_lock = start.elapsed();
+        println!("{}:{}", file!(), line!());
```
Comment (hansieodendaal) [inline base_layer/wallet/src/output_manager_service/storage/sqlite_db/mod.rs]: Was print this perhaps for debugging?

File: base_layer/wallet/src/output_manager_service/storage/sqlite_db/mod.rs [hunk 6]
```diff
@@ -756,8 +756,10 @@ impl OutputManagerBackend for OutputManagerSqliteDatabase {
...
 
         let outputs_to_be_received = OutputSql::index_status(OutputStatus::ShortTermEncumberedToBeReceived, &conn)?;
+        println!("{}:{}", file!(), line!());
```
Comment (hansieodendaal) [inline base_layer/wallet/src/output_manager_service/storage/sqlite_db/mod.rs]: Print?

File: base_layer/wallet/src/output_manager_service/storage/sqlite_db/mod.rs [hunk 7]
```diff
@@ -768,7 +770,9 @@ impl OutputManagerBackend for OutputManagerSqliteDatabase {
             )?;
         }
 
+        println!("{}:{}", file!(), line!());
```
Comment (hansieodendaal) [inline base_layer/wallet/src/output_manager_service/storage/sqlite_db/mod.rs]: Print?

File: base_layer/wallet/src/output_manager_service/storage/sqlite_db/mod.rs [hunk 8]
```diff
@@ -768,7 +770,9 @@ impl OutputManagerBackend for OutputManagerSqliteDatabase {
...
         let outputs_to_be_spent = OutputSql::index_status(OutputStatus::ShortTermEncumberedToBeSpent, &conn)?;
+        println!("{}:{}", file!(), line!());
```
Comment (hansieodendaal) [inline base_layer/wallet/src/output_manager_service/storage/sqlite_db/mod.rs]: Print?

File: base_layer/wallet/src/output_manager_service/storage/sqlite_db/mod.rs [hunk 9]
```diff
@@ -787,6 +791,7 @@ impl OutputManagerBackend for OutputManagerSqliteDatabase {
                 start.elapsed().as_millis()
             );
         }
+        println!("{}:{}", file!(), line!());
```
Comment (hansieodendaal) [inline base_layer/wallet/src/output_manager_service/storage/sqlite_db/mod.rs]: Print?

File: base_layer/core/src/transactions/transaction/output_features.rs [hunk 10]
```diff
@@ -43,9 +44,26 @@ use crate::{
     },
 };
 
+#[derive(Debug, Clone, Hash, PartialEq, Deserialize, Serialize, Eq)]
+#[repr(u8)]
+pub enum OutputFeaturesVersion {
+    DIBBLER = 0,
+}
+
```
Comment (hansieodendaal) [inline base_layer/core/src/transactions/transaction/output_features.rs]: I think we need an `OutputFeaturesVersion::GetCurrentVersion()` implementation here.

File: base_layer/core/src/transactions/transaction/output_features.rs [hunk 11]
```diff
@@ -63,6 +81,53 @@ impl OutputFeatures {
...
+        maturity: u64,
+        metadata: Vec<u8>,
+        unique_id: Option<Vec<u8>>,
+        parent_public_key: Option<PublicKey>,
+        asset: Option<AssetOutputFeatures>,
+        mint_non_fungible: Option<MintNonFungibleFeatures>,
+        sidechain_checkpoint: Option<SideChainCheckpointFeatures>,
+    ) -> OutputFeatures {
+        OutputFeatures::new(
+            OutputFeaturesVersion::DIBBLER,
```
Comment (hansieodendaal) [inline base_layer/core/src/transactions/transaction/output_features.rs]:
```suggestion
            OutputFeaturesVersion::GetCurrentVersion(),
```

File: base_layer/core/src/transactions/transaction/transaction_input.rs [hunk 12]
```diff
@@ -56,11 +58,29 @@ use crate::{
     },
 };
 
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
+#[repr(u8)]
+pub enum TransactionInputVersion {
+    DIBBLER = 0,
+}
+
```
Comment (hansieodendaal) [inline base_layer/core/src/transactions/transaction/transaction_input.rs]: I think we need a `TransactionInputVersion::GetCurrentVersion()` implementation here.

File: base_layer/core/src/transactions/transaction/transaction_input.rs [hunk 13]
```diff
@@ -71,19 +91,42 @@ pub struct TransactionInput {
...
         }
     }
 
+    pub fn new_current_version(
+        spent_output: SpentOutput,
+        input_data: ExecutionStack,
+        script_signature: ComSignature,
+    ) -> TransactionInput {
+        TransactionInput::new(
+            TransactionInputVersion::DIBBLER,
```
Comment (hansieodendaal) [inline base_layer/core/src/transactions/transaction/transaction_input.rs]:
```suggestion
            TransactionInputVersion::GetCurrentVersion(),
```

File: base_layer/core/src/transactions/transaction/transaction_input.rs [hunk 14]
```diff
@@ -94,17 +137,18 @@ impl TransactionInput {
...
-        TransactionInput {
-            spent_output: SpentOutput::OutputData {
+        TransactionInput::new_current_version(
+            SpentOutput::OutputData {
                 features,
                 commitment,
                 script,
                 sender_offset_public_key,
                 covenant,
+                version: TransactionOutputVersion::DIBBLER,
```
Comment (hansieodendaal) [inline base_layer/core/src/transactions/transaction/transaction_input.rs]:
```suggestion
                version: TransactionOutputVersion::GetCurrentVersion(),
```

File: base_layer/core/src/transactions/transaction/transaction_kernel.rs [hunk 15]
```diff
@@ -62,6 +80,41 @@ pub struct TransactionKernel {
...
+
+    pub fn new_current_version(
+        features: KernelFeatures,
+        fee: MicroTari,
+        lock_height: u64,
+        excess: Commitment,
+        excess_sig: Signature,
+    ) -> TransactionKernel {
+        TransactionKernel::new(
+            TransactionKernelVersion::DIBBLER,
```
Comment (hansieodendaal) [inline base_layer/core/src/transactions/transaction/transaction_kernel.rs]:
```suggestion
            TransactionKernelVersion::GetCurrentVersion(),
```

File: base_layer/core/src/transactions/transaction/transaction_output.rs [hunk 16]
```diff
@@ -115,6 +136,27 @@ impl TransactionOutput {
...
+        features: OutputFeatures,
+        commitment: Commitment,
+        proof: RangeProof,
+        script: TariScript,
+        sender_offset_public_key: PublicKey,
+        metadata_signature: ComSignature,
+        covenant: Covenant,
+    ) -> TransactionOutput {
+        TransactionOutput::new(
+            TransactionOutputVersion::DIBBLER,
```
Comment (hansieodendaal) [inline base_layer/core/src/transactions/transaction/transaction_output.rs]:
```suggestion
            TransactionOutputVersion::GetCurrentVersion(),
```

File: base_layer/core/src/transactions/transaction/unblinded_output.rs [hunk 17]
```diff
@@ -99,6 +103,33 @@ impl UnblindedOutput {
...
+        script: TariScript,
+        input_data: ExecutionStack,
+        script_private_key: PrivateKey,
+        sender_offset_public_key: PublicKey,
+        metadata_signature: ComSignature,
+        script_lock_height: u64,
+        covenant: Covenant,
+    ) -> Self {
+        Self::new(
+            TransactionOutputVersion::DIBBLER,
```
Comment (hansieodendaal) [inline base_layer/core/src/transactions/transaction/unblinded_output.rs]:
```suggestion
            TransactionOutputVersion::GetCurrentVersion(),
```

File: base_layer/core/src/transactions/transaction/transaction_kernel.rs [hunk 18]
```diff
@@ -39,13 +40,30 @@ use crate::transactions::{
     transaction_protocol::{build_challenge, TransactionMetadata},
 };
 
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
+#[repr(u8)]
+pub enum TransactionKernelVersion {
+    DIBBLER = 0,
+}
+
```
Comment (hansieodendaal) [inline base_layer/core/src/transactions/transaction/transaction_kernel.rs]: I think we need a `TransactionKernelVersion::GetCurrentVersion()` implementation here.

File: base_layer/core/src/transactions/transaction/transaction_output.rs [hunk 19]
```diff
@@ -70,11 +71,29 @@ use crate::{
     },
 };
 
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
+#[repr(u8)]
+pub enum TransactionOutputVersion {
+    DIBBLER = 0,
+}
+
```
Comment (hansieodendaal) [inline base_layer/core/src/transactions/transaction/transaction_output.rs]: I think we need a `TransactionOutputVersion::GetCurrentVersion()` implementation here.

**Review by hansieodendaal — COMMENTED** [review]
Hi there,

Just some comments to consider.

Thanks.

File: base_layer/core/src/transactions/transaction/output_features.rs [hunk 20]
```diff
@@ -43,9 +44,26 @@ use crate::{
     },
 };
 
+#[derive(Debug, Clone, Hash, PartialEq, Deserialize, Serialize, Eq)]
+#[repr(u8)]
+pub enum OutputFeaturesVersion {
+    DIBBLER = 0,
```
Comment (sdbondi) [inline base_layer/core/src/transactions/transaction/output_features.rs]: Versions don't correspond with networks. Suggest `V1`, `V2` etc

File: applications/tari_app_grpc/src/conversions/transaction_output.rs [hunk 21]
```diff
@@ -56,15 +56,16 @@ impl TryFrom<grpc::TransactionOutput> for TransactionOutput {
             .try_into()
             .map_err(|_| "Metadata signature could not be converted".to_string())?;
         let covenant = Covenant::consensus_decode(&mut output.covenant.as_slice()).map_err(|err| err.to_string())?;
-        Ok(Self {
+        Ok(Self::new(
+            TransactionOutputVersion::try_from(output.version as u8)?,
```
Comment (sdbondi) [inline applications/tari_app_grpc/src/conversions/transaction_output.rs]: Should check for u8 overflows
```suggestion
            TransactionOutputVersion::try_from(u8::try_from(output.version).map_err(|_| "Invalid version: overflowed u8"))?,
```

File: base_layer/wallet/src/output_manager_service/storage/sqlite_db/mod.rs (hunk 7 above)
Comment (Cifko) [reply base_layer/wallet/src/output_manager_service/storage/sqlite_db/mod.rs]: Forgot some debug ;-)

File: base_layer/core/src/transactions/transaction/output_features_version.rs [hunk 22]
```diff
@@ -0,0 +1,26 @@
+use std::convert::TryFrom;
+
+use serde::{Deserialize, Serialize};
+
+#[derive(Debug, Clone, Hash, PartialEq, Deserialize, Serialize, Eq)]
+#[repr(u8)]
+pub enum OutputFeaturesVersion {
+    V1 = 0,
```
Comment (stringhandler) [inline base_layer/core/src/transactions/transaction/output_features_version.rs]:
```suggestion
    V0 = 0,
```

File: base_layer/core/src/transactions/transaction/transaction_input_version.rs [hunk 23]
```diff
@@ -0,0 +1,26 @@
+use std::convert::TryFrom;
+
+use serde::{Deserialize, Serialize};
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
+#[repr(u8)]
+pub enum TransactionInputVersion {
+    V1 = 0,
```
Comment (stringhandler) [inline base_layer/core/src/transactions/transaction/transaction_input_version.rs]:
```suggestion
    V0 = 0,
```

File: base_layer/core/src/transactions/transaction/transaction_kernel_version.rs [hunk 24]
```diff
@@ -0,0 +1,25 @@
+use std::convert::TryFrom;
+
+use serde::{Deserialize, Serialize};
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
+#[repr(u8)]
+pub enum TransactionKernelVersion {
+    V1 = 0,
```
Comment (stringhandler) [inline base_layer/core/src/transactions/transaction/transaction_kernel_version.rs]:
```suggestion
    V0 = 0,
```

File: base_layer/core/src/blocks/genesis_block.rs [hunk 25]
```diff
@@ -219,29 +219,29 @@ fn get_dibbler_genesis_block_raw() -> Block {
     );
     let mut body = AggregateBody::new(
         vec![],
-        vec![TransactionOutput {
-            features: OutputFeatures {
+        vec![TransactionOutput::new_current_version(
```
Comment (SWvheerden) [inline base_layer/core/src/blocks/genesis_block.rs:222]: You have changed the hash of the output, you need to update the MMR in the header of the genesis block as well.

File: base_layer/core/src/transactions/transaction/mod.rs [hunk 26]
```diff
@@ -92,13 +100,15 @@ pub(super) fn hash_output(
...
     covenant: &Covenant,
+    version: TransactionOutputVersion,
 ) -> Vec<u8> {
     HashDigest::new()
         .chain(features.to_consensus_bytes())
         .chain(commitment.as_bytes())
         // .chain(range proof) // See docs as to why we exclude this
         .chain(script.as_bytes())
         .chain(covenant.to_consensus_bytes())
+        .chain((version as u8).to_le_bytes())
```
Comment (SWvheerden) [inline base_layer/core/src/transactions/transaction/mod.rs:111]: Look at the comment in the genesis block, you need to update the MMR in the header.

**Review by SWvheerden — CHANGES_REQUESTED** [review]
This looks good, but you need to remember to update the genesis block header as well as the MMR has now changed.

**Comment by Cifko** [comment]
I've also forgot to add version to the TransactionInput hash. Added that as well. The genesis block is updated.

**Comment by Cifko** [comment]
Added `Clippy.toml` for the number of arguments (set to 15) and removed the `#[allow(clippy::too_many_arguments)]`. This is only done for the `base_layer/core` crate.

**Review by stringhandler — APPROVED** [review]

**Pull request merged**